applied bootstack-charms-spec template for snaps

gabrielcocenza opened the pull request:
- added unit and functional tests
- used uvicorn + fastapi for http server

gabrielcocenza committed:
applied bootstack-charms-spec template for snaps

- added unit and functional tests
- used uvicorn + fastapi for http server

mkalcok commented on software_inventory_exporter/exporter.py:
```diff
@@ -0,0 +1,132 @@
+        return output
+    except subprocess.TimeoutExpired as error:
+        logger.error("Timeout to list snap %s", error)
+        raise HTTPException(status_code=500, detail="Timeout error")
```
Returning `Timeout error` with response code `500` is confusing imo, it suggests that the request timed out, when in fact it's an internal server error.

I think that all `500` errors should return the same generic error message and details about the error should be stored in logs exclusively.

mkalcok reviewed (changes requested):
Few small suggestions, otherwise LGTM.

gabrielcocenza committed:
- used argparse instead of sys.argv.
- api.py -> contains fastapi endpoints.
- exporter.py -> generate content for the endpoint and handle
  exceptions.
- cli.py -> entrypoint of the software.

gabrielcocenza commented:
Thanks for your review @mkalcok.
I've updated the patch with your suggestions.

mkalcok reviewed (approved):
thanks for the updates. LGTM, +1

gabrielcocenza committed:
added CODEOWNERS file in the project.

gabrielcocenza committed:
- added the github workflows template for snaps.

rgildein commented on tests/functional/conftest.py:
```diff
@@ -0,0 +1,55 @@
+        install_cmd = (
+            f"sudo snap install --dangerous --classic {software_inventory_exporter_snap}".split()
+        )
+        os.chdir("/tmp")
```
why we need to change directory to '/tmp' here?

gabrielcocenza replied:
I had a issue with `juju-lint` snap (also classic) that if I don't change the working directory, it will run the modules instead of running the snap. Python will have a preference running the code thru the modules and sometimes that can be misleading. E.g: My setup.py was not including non-python files and the functional tests were passing and it wasn't suppose to pass.

Pjack replied:
Seems wired. But if it is true, I suggest to add the comment to explain it.

rgildein replied:
I see because venv is activated and instead of `/snap/bin/juju-lint` it used python version. I that changing the directory to tmp here is maybe not the best approach, but I could not see any other now. Can you put some comments to hat, so if someone will came he/she will know that why it's done. Thanks

rgildein replied:
This is not-blocking

gabrielcocenza replied:
Line 21 of this file already has a comment why I'm changing the working directory.

rgildein commented on tox.ini:
```diff
@@ -0,0 +1,75 @@
+    -r {toxinidir}/tests/unit/requirements.txt
+    pytest
+    pytest-cov
+setenv = PYTHONPATH={toxinidir}
```
This is not current version, can you check all templates again to be in current version.

rgildein reviewed (changes requested):
I think you need to checkout to current master branch for bootstack-charms-spec and update all templates here.
I'll also suggest to make two separate PR next time. One with templates and second with code itself.

mkalcok commented on software_inventory_exporter/cli.py:
```diff
@@ -0,0 +1,41 @@
+
+    try:
+        config = yaml.safe_load(target_config.read_text(encoding="utf-8"))["settings"]
+        uvicorn.run(app, host=config["bind_address"], port=int(config["port"]))
```
No, service upkeep (run/stop/restart/autorestart) is handled by systemd.

gabrielcocenza commented on software_inventory_exporter/exporter.py:
```diff
@@ -0,0 +1,88 @@
+    try:
+        dpkg = str(subprocess.check_output(cmd.split(), timeout=10))
+        lines = dpkg.split("\\n")
+        lines = lines[5:-1]
```
I've showed what is expected as output in the unit tests, but this is a good idea.
Thanks

gabrielcocenza committed:
- updated from newest bootstack-spec templates
